From 934895a8e5ff2645f659285db1d11198a9167dd5 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Thu, 14 Aug 2025 01:28:38 +0200
Subject: [PATCH 55/98] Settings: Add lineagehealth IFastCharge interface
 support

Create a new charging category to blend with the new UI.

Co-authored-by: Bruno Martins <bgcngm@gmail.com>
Change-Id: If5b88efde3bf71e6c38ee0e6ac812ed1a3287b13
---
 res/xml/power_usage_summary.xml               |  21 +++-
 .../ChargingSpeedPreferenceController.java    | 106 ++++++++++++++++++
 2 files changed, 123 insertions(+), 4 deletions(-)
 create mode 100644 src/com/android/settings/fuelgauge/ChargingSpeedPreferenceController.java

diff --git a/res/xml/power_usage_summary.xml b/res/xml/power_usage_summary.xml
index 7e34dd3b091..663d82fbf50 100644
--- a/res/xml/power_usage_summary.xml
+++ b/res/xml/power_usage_summary.xml
@@ -74,10 +74,23 @@
             settings:requiresConfigMask="64" />
     </com.android.settingslib.widget.UntitledPreferenceCategory>
 
-    <!-- Charging control -->
-    <lineageos.preference.LineagePartsPreference
-        android:key="charging_control"
-        settings:requiresService="lineagehealth" />
+    <com.android.settingslib.widget.UntitledPreferenceCategory
+        android:key="charging_category">
+        <!-- Charging control -->
+        <lineageos.preference.LineagePartsPreference
+            android:key="charging_control"
+            settings:requiresService="lineagehealth" />
+
+        <!-- Charging speed -->
+        <ListPreference
+            android:entries="@*lineageos.platform:array/charging_speed_entries"
+            android:entryValues="@*lineageos.platform:array/charging_speed_values"
+            android:key="charging_speed"
+            android:summary="%s"
+            android:title="@*lineageos.platform:string/charging_speed"
+            settings:controller="com.android.settings.fuelgauge.ChargingSpeedPreferenceController"
+            settings:requiresService="lineagehealth" />
+    </com.android.settingslib.widget.UntitledPreferenceCategory>
 
     <com.android.settingslib.widget.UntitledPreferenceCategory
         android:key="percentage_category">
diff --git a/src/com/android/settings/fuelgauge/ChargingSpeedPreferenceController.java b/src/com/android/settings/fuelgauge/ChargingSpeedPreferenceController.java
new file mode 100644
index 00000000000..a120ca62dd2
--- /dev/null
+++ b/src/com/android/settings/fuelgauge/ChargingSpeedPreferenceController.java
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2025 The LineageOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.fuelgauge;
+
+import android.content.Context;
+
+import androidx.preference.ListPreference;
+import androidx.preference.Preference;
+import androidx.preference.PreferenceScreen;
+
+import com.android.internal.util.ArrayUtils;
+import com.android.settings.core.BasePreferenceController;
+
+import lineageos.health.HealthInterface;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Controller to change and update the fast charging settings
+ */
+public class ChargingSpeedPreferenceController extends BasePreferenceController
+        implements Preference.OnPreferenceChangeListener {
+
+    private HealthInterface mHealthInterface;
+    private ListPreference mListPreference;
+
+    public ChargingSpeedPreferenceController(Context context, String key) {
+        super(context, key);
+
+        try {
+            mHealthInterface = HealthInterface.getInstance(context);
+        } catch (RuntimeException e) {
+            // Not supported.
+        }
+    }
+
+    @Override
+    public void displayPreference(PreferenceScreen screen) {
+        super.displayPreference(screen);
+
+        mListPreference = screen.findPreference(getPreferenceKey());
+
+        if (getAvailabilityStatus() == UNSUPPORTED_ON_DEVICE) {
+            return;
+        }
+
+        List<CharSequence> entries = new ArrayList<>();
+        entries.addAll(Arrays.asList(mListPreference.getEntries()));
+
+        List<CharSequence> entryValues = new ArrayList<>();
+        entryValues.addAll(Arrays.asList(mListPreference.getEntryValues()));
+
+        int[] supportedFastChargeModes = mHealthInterface.getSupportedFastChargeModes();
+
+        for (int i = entryValues.size() - 1; i >= 0; i--) {
+            int fastChargeMode = Integer.parseInt(entryValues.get(i).toString());
+            if (!ArrayUtils.contains(supportedFastChargeModes, fastChargeMode)) {
+                entries.remove(i);
+                entryValues.remove(i);
+            }
+        }
+
+        mListPreference.setEntries(entries.stream().toArray(CharSequence[]::new));
+        mListPreference.setEntryValues(entryValues.stream().toArray(CharSequence[]::new));
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return mHealthInterface != null && mHealthInterface.isFastChargeSupported()
+                ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        String fastChargeMode = String.valueOf(mHealthInterface.getFastChargeMode());
+        int index = mListPreference.findIndexOfValue(fastChargeMode);
+        if (index < 0) index = 0;
+        mListPreference.setValueIndex(index);
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        if (mHealthInterface.setFastChargeMode(Integer.parseInt((String) newValue))) {
+            updateState(preference);
+            return true;
+        }
+
+        return false;
+    }
+}
